book/src/pong_tutorial/pong_tutorial_02.md

```diff
@@ -0,0 +1,367 @@
+fn initialise_paddles(world: &mut World) {
+
+    let mut left_transform = LocalTransform::default();
+    let mut right_transform = LocalTransform::default();
+
+    // Correctly position the paddles.
+    let y = -PADDLE_HEIGHT / 2.0;
+    left_transform.translation = Vector3::new(-1.0, y, 0.0);
+    right_transform.translation = Vector3::new(1.0 - PADDLE_WIDTH, y, 0.0);
+
+
+    // Create a left plank entity.
+    world
+        .create_entity()
+        .with(Paddle::new(Side::Left))
+        .with(Transform::default())
+        .with(left_transform)
+        .build();
+
+    // Create right plank entity.
+    world
+        .create_entity()
+        .with(Paddle::new(Side::Right))
+        .with(Transform::default())
+        .with(right_transform)
+        .build();
+}
+```
+
+This is all the information Amethyst needs to track and move the paddles in our 
+virtual world, but we'll need to do some more work to actually *draw* them.
+
+## Drawing
+
+Here's a utility function to generate the six vertices of a rectangle (a 
+rectangle in computer graphics is typically drawn with two triangles):
+
+```rust,ignore
+fn generate_rectangle_vertices(left: f32,
+                               bottom: f32,
+                               right: f32,
+                               top: f32) -> Vec<PosTex> {
+    vec![
+        PosTex {
+            position: [left, bottom, 0.],
+            tex_coord: [0.0, 0.0],
+        },
+        PosTex {
+            position: [right, bottom, 0.0],
+            tex_coord: [1.0, 0.0],
+        },
+        PosTex {
+            position: [left, top, 0.0],
+            tex_coord: [1.0, 1.0],
+        },
+        PosTex {
+            position: [right, top, 0.],
+            tex_coord: [1.0, 1.0],
+        },
+        PosTex {
+            position: [left, top, 0.],
+            tex_coord: [0.0, 1.0],
+        },
+        PosTex {
+            position: [right, bottom, 0.0],
+            tex_coord: [0.0, 0.0],
+        },
+    ]
+}
+```
+
+`PosTex` is a type defined by `amethyst_renderer`. It's a vertex format with 
+position and UV texture coordinate attributes. In our rendering pipeline, if 
+you'll recall, we created a `DrawFlat::<PosTex>` pass, which draws a `PosTex` 
+mesh. Right now our vertices are simply in a standard Rust `Vector`. To create a 
+mesh from them we'll write another utility function, which generates a 
+`MeshHandle`:
+
+```rust,ignore
+fn create_mesh(world: &World, vertices: Vec<PosTex>) -> MeshHandle {
+    let loader = world.read_resource::<Loader>();
+    loader.load_from_data(vertices.into(), (), &world.read_resource())
+}
+```
+
+The `Loader` is an asset loader which is defined as a `resource` (not a `Entity`
+, `Component`, or `System`, but still a part of our ECS `world`). It was created 
+when we built our Application in `main.rs`, and it can read assets like .obj 
+files, but also it can `load_from_data` as in our use case.
```

Comment:
An elaboration on what `data` means in terms of the Loader might be helpful here for new readers.  The asset system is kind of it's own beast in terms of operation so an in depth explanation here of how to use it would be very good.

I'm happy to provide this explanation if need be, @torkleyy is also a good resource since he wrote the thing.

Reply:
Yeah I'd love more information here, and what I wrote is pretty much the limit of my understanding right now. Took me a while to figure out how we even had a Loader to call.

Reply:
You can pick and choose what's necessary from this explanation, but I'll give you the whole rundown so you know for future reference.

The asset loading system is designed to interpret data from the hard drive asynchronously.  This asynchronous management is necessary because I/O is slower than is tolerable for a videogame frame.  First a `Loader` is needed, one of these is created and added to the world as a resource when the application is built.  When you want to load assets you'll need to get the resource and call one of the `load*` functions on it.

- `load` This function reads from files in the default directory provided at startup.
  - name: this is the file path relative to the default directory.
  - format: A structure implementing the Format trait, the Format trait implementation for the structure describes how to interpret the raw binary data as an asset that's useful to the engine.  `Asset`s are any structure implementing the `Asset` trait.
  - options: If a format needs special instructions for how to interpret a file you'll pass them here, otherwise you'll pass `()`
  - progress: A structure that receives progress information from the `Loader`, useful for programming progress bars when waiting for a level to load.  Alternatively, if you don't need progress information you can just use `()` here.
  - storage: The storage to load the asset into.  This is typically also a resource.

- `load_from` is very similar to `load` but instead of loading from the default directory it will load from the directory given in the `source` argument.

- `load_from_data` the `Asset` trait also provides an associated type called `Data` when implemented.  If you don't want to load an asset from disk and are able to programmatically produce a structure of the `Data` type, all you have to do to load it is supply it to this function.

All of these functions return `Handle`s to the asset being loaded.  Later on when you need the asset you can request the asset from the `AssetStorage` with your `Handle`, which will return the asset if it has finished loading the asset.

Alternatively, many Asset types are typically attached directly to an entity when they're loaded, so the `Bundle`s that register those `AssetStorages` will also typically register an asset `Processor` system.  Since `Handle` implements `Component` if you attach the `Handle` to an entity then later when the asset is ready the `Processor` system will then add the loaded asset to the Entity with the `Handle` component.

So what's going in the pong example is that the mesh `Data` is prepared and then fed into `load_from_data` which will return a `Handle<Mesh>` also known as a `MeshHandle`.  Then we attach this handle to our entity and when the asset has finished loading it will be attached to the Entity in question.  It's worth noting in this particular instance that because there is no disk I/O the asset will simply be attached next frame.

Reply:
The extra step for the `Mesh` attachment in the pong example, is because a mesh isn't done loading after the data is available, it also needs to be uploaded to graphics memory, which is the cause of `attached next frame` in @Xaeroxe's explanation.

Reply:
> the `Processor` system will then add the loaded asset to the Entity with the Handle component

I'm not sure what you're talking about, are you confusing this with the old asset loading?

Reply:
@torkleyy I'm referencing this: https://www.amethyst.rs/doc/develop/doc/amethyst_assets/struct.Processor.html and any systems designed to fulfill the same purpose.